### DISCUSSION TITLE
qt6 vulkan

### DISCUSSION THREAD

**Discussion by jQill** [OE4T/meta-tegra#1448] [Q&A]
I am currently building a Yocto image for the Jetson Nano that runs my Qt application in kiosk mode, utilizing the OpenGL EGLFS kms/drm backend. Recently, I came across information in the Qt 6 documentation about its Vulkan support. I made several attempts to build the image, but from what I understand, it requires a desktop environment like X11 or Wayland. Is that correct? Perhaps there is a way to build it without involving graphic components like these?

**Comment by EasyIP2023** [comment]
@jQill not that familiar with QT6, but vulkan-kms support is more than likely not built into QT yet.
I'd see it anyone is attempting to support it there first.

QT would have to implement a number of Vulkan logical device extensions

VK_KHR_external_memory_fd
VK_KHR_external_semaphore_fd
VK_KHR_synchronization2
VK_KHR_image_format_list
VK_EXT_external_memory_dma_buf
VK_EXT_queue_family_foreign
VK_EXT_image_drm_format_modifier

To name a few

QT would have to place Vulkan pipeline rendered frames directly into GEM buffers. To do so export those GEM buffers
as DMA-BUFs. You can create GEM buffers with gbm and ioctl calls. Nvidia doesn't support gbm (far as I know), but I'd leave
@kekiefer to answer that as he was working on a wrapper for gbm I believe.

wlroots is an excellent resource to try and see if you can get the Vulkan renderer working with drm backend on Jetson line up.

https://gitlab.freedesktop.org/wlroots/wlroots

I've been working on a side project for a little less than 2 years now (when I have free time). It's no where near on the level of wlroots,
but I at least see frames. Can try to get it going on tegra if you'd like. I usually test with AMDGPU's or on the udoo-bolt.

https://github.com/under-view/kmsroots/tree/master

**Comment by kekiefer** [comment]
GBM needs just a little help to bridge the gbm bo to nvbuffer interfaces with a [custom gbm backend](https://github.com/OE4T/tegra-udrm-gbm), and then Qt's eglfs-kms-gbm works as well as weston and kmscube at least.

Set `PREFERRED_PROVIDER_tegra-gbm-backend = "tegra-udrm-gbm"` to use this, and that should take care of the display interface side of your problem.

**Reply by jQill** [reply]
@kekiefer
Hm, I am not deeply involved in this and do not completely understand. What exact problem does this flag and repository solve? Does Eglfs start using the NVBuffer memory API instead of system memory? It may be unclear from the main post, but I succeeded in running a Qt application with Eglfs like this:
QT_QPA_PLATFORM=eglfs QT_QPA_EGLFS_INTEGRATION=eglfs_kms_egldevice ./appqt6test
However, this solution has a major issue: it consumes a lot of system memory, at least 0.5 GB for a simple rendering demo application that anyone can create using Qt Creator (it was built with optimizations, of course). And I was wondering if using Vulkan instead of EGL will help me solve this.
